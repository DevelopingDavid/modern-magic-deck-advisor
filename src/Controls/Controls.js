import React, {Component} from 'react';
import '../styles/main.scss';

class Controls extends Component {
  constructor(props) {
    super(props);
    this.textInput = React.createRef();
    this.state = {
      cardName: '',
      isInvalidCardName: false,
      hasDuplicates: false,
      suggestions: []
    };
  }

  updateCardName = (event) => {
    let newSuggestions = [];
    if (event.target.value.length !== 0) {
      newSuggestions = this.props.cards.filter(card => {
        const cardName = card.cardName.toLowerCase();
        return cardName.startsWith(event.target.value.toLowerCase());
      })
        .map(card => card.cardName);
    }
    this.setState({ 
      cardName: event.target.value,
      suggestions: newSuggestions
     })
  }

  chooseSuggestion = (event) => {
    this.textInput.current.value = event.target.id;
    this.textInput.current.focus();
    this.setState({
      cardName: event.target.id,
      suggestions: [] 
    });
  }

  submitCard = (event) => {
    event.preventDefault();
    const matchedCard = this.validateCardName();
    if (matchedCard) {
      this.addCardToList(matchedCard.cardName)
    } else {
      this.throwInvalidCardNameError();
    }
    this.clearInput();
  }

  validateCardName = () => {
    return this.props.cards.find(card => {
      return card.cardName.toLowerCase() === this.state.cardName.toLowerCase();
    });
  }

  addCardToList = (cardName) => {
    let cardNames = this.props.userCardsData.map(card => card.cardName);
    if (!cardNames.includes(cardName)) {
      this.setState({
        hasDuplicates: false,
        isInvalidCardName: false
      });
      this.props.addUserCard(cardName);
    } else {
      this.setState({ 
        hasDuplicates: true,
        isInvalidCardName: false
      });
    }
  }

  throwInvalidCardNameError = () => {
    this.setState({ isInvalidCardName: true });
  }

  clearInput = () => {
    this.textInput.current.value = '';
    this.textInput.current.focus();
  }

  render() {
    let errorMessage;
    if (this.state.isInvalidCardName) {
      errorMessage = "The card name you entered is not in the database"
    } else if (this.state.hasDuplicates) {
      errorMessage = "This card is already in your list"
    }
    return(
      <form>
        <input
          type="text"
          className="controls--input"
          onChange={this.updateCardName}
          placeholder="Enter cards you own"
          ref={this.textInput}>
        </input>
        <button
          className="controls--button"
          onClick={this.submitCard}>
            Add Card
        </button>
        <div className="controls--suggestion-area">
          {this.state.suggestions.map(cardName => {
            return (
              <span
                className="controls--suggestion"
                key={cardName}
                id={cardName}
                onClick={this.chooseSuggestion}>
              {cardName.slice(0,24)}
              </span>
            );
          })}
        </div>
        <p className="controls--error">{errorMessage}</p>
      </form>
    )
  }
}

export default Controls;