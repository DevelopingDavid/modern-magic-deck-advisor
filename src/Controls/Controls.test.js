import React from 'react';
import Controls from './Controls';
import { shallow } from 'enzyme';
import cards from '../mockData';

const addCardToListMock = jest.fn();
const throwInvalidCardNameErrorMock = jest.fn();

describe('Controls', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Controls
        addCardToList={addCardToListMock}
        throwInvalidCardNameError={throwInvalidCardNameErrorMock}
        isInvalidCardName={false}
        hasDuplicates={false}
        cards={cards}/>
    )
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should should set the state of cardName on change of the input', () => {
    wrapper.find('.controls--input').simulate('change', { target: { value: 'B' } });
    expect(wrapper.state()).toEqual({ cardName: 'B' });
  })
  
  it('should call props.addCardToList when the button is clicked and there is a matchedCard', () => {
    wrapper.find('.controls--input').simulate('change', { target: { value: 'Blood Moon' } });
    wrapper.find('.controls--button').simulate('click', { preventDefault: () => {} });
    expect(addCardToListMock).toBeCalled();
  })
  
  it('should call props.throwInvalidCardNameError when the button is clicked and there is no matchedCard', () => {
    wrapper.find('.controls--input').simulate('change', { target: { value: 'asdfgasd' } });
    wrapper.find('.controls--button').simulate('click', { preventDefault: () => {} });
    expect(throwInvalidCardNameErrorMock).toBeCalled();
  });
  
  it('should return a card object if the state.cardName matches a card in the data set when validateCardName is called', () => {
    wrapper.find('.controls--input').simulate('change', { target: { value: 'Tarmogoyf' } });
    expect(wrapper.instance().validateCardName().cardName).toEqual('Tarmogoyf');
  })
})