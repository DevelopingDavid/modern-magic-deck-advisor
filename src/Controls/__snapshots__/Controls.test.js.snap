// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls should match the snapshot 1`] = `
<form>
  <input
    className="controls--input"
    onChange={[Function]}
    type="text"
  />
  <button
    className="controls--button"
    onClick={[Function]}
  >
    Click to add card
  </button>
  <p />
</form>
`;
